mongo/options/clientoptions.go

```diff
@@ -729,7 +729,8 @@ func (c *ClientOptions) SetSocketTimeout(d time.Duration) *ClientOptions {
 // (e.g. "timeoutMS=1000"). The default value is nil, meaning operations do not inherit a timeout from the Client.
 //
 // If any Timeout is set (even 0) on the Client, the values of other, deprecated timeout-related options will be ignored.
-// In particular: ClientOptions.SocketTimeout, WriteConcern.wTimeout, MaxTime on operations, and TransactionOptions.MaxCommitTime.
+// In particular: ClientOptions.SocketTimeout, WriteConcern.wTimeout, MaxTime on operation options, and
+// TransactionOptions.MaxCommitTime.
```

Comment:
Implementing the "ignoring" behavior for legacy timeout options described here seems to introduce a considerable amount of complexity for a small possible benefit. Most timeout options are not technically incompatible with `timeoutMS` (i.e. do not introduce conflicting values of the same timeout setting), although using them together is not advisable. The main exception is `maxTimeMS`, which does potentially conflict when both a client-level `timeoutMS` and operation-level `maxTimeMS` are set.

Does the usability benefit of ignoring legacy timeout options outweigh the increased driver complexity? Can we more simply accomplish a similar outcome by documenting that `timeoutMS` should not be used with `SocketTimeout`, `WriteConcern.wTimeout`, etc?

Reply:
An interesting question; I think the idea of solely deprecating and not having `timeoutMS` override the deprecated options was [considered](https://github.com/mongodb/specifications/blob/master/source/client-side-operations-timeout/client-side-operations-timeout.rst#timeoutms-overrides-deprecated-timeout-options) in the specification design. The design I have here to ignore the timeouts does not seem considerably complex IMO; most of the code changes in this PR have to do with centralizing the calculation of `maxTimeMS`. Do you think this ignoring behavior will be difficult to maintain in the future?

Also, I think every deprecated option _could_ feasibly conflict with `timeoutMS`. A high `SocketTimeout` may not be honored if a low `Timeout` causes a context deadline exceeded error before `SocketTimeout` is reached. A high `wTimeout` or `MaxCommitTime`  not be honored if a low `Timeout` causes a `MaxTimeMSExpired` error with its automatic `maxTimeMS` appension. And, as you say, `Timeout` and `MaxTime` will directly conflict.

Reply:
I've brought this up offline with @ShaneHarvey. I think I agree that the ignoring behavior is not necessarily less confusing for users and adds code complexity.